WZDx v4.0

j-d-b opened the pull request:
# WZDx v4.0

WZDx version 4.0 implements clean up and small additions in functionality to the WZDx feed and adds definitions for two new feeds, the [SwzDeviceFeed](/spec-content/objects/SwzDeviceFeed.md) and [RoadRestrictionFeed](/spec-content/objects/RoadRestrictionFeed.md). Until version 4.0, the WZDx specification defined only one feed, the [WZDxFeed](/spec-content/objects/WZDxFeed.md).

## Features
- Add the following values to the [VehicleImpact](/spec-content/enumerated-types/VehicleImpact.md) enumerated type:
    - `some-lanes-closed-merge-left`
    - `some-lanes-closed-merge-right`
    - `all-lanes-open-shift-left`
    - `all-lanes-open-shift-right`
    - `some-lanes-closed-split`
    - `flagging`
    - `temporary-traffic-signal`
- Allow restrictions with a value and unit to be provided at the road event level; specifically:
    - Rename the LaneRestrictionUnit enumerated type to [UnitOfMeasurement](/spec-content/enumerated-types/UnitOfMeasurement.md).
    - Rename the RoadRestriction enumerated type to [RestrictionType](/spec-content/enumerated-types/RestrictionType.md).
    - Rename the LaneRestriction object to [Restriction](/spec-content/objects/Restriction.md) and remove the `lane_restriction_` prefix from its properties.
    - Rename the [Restriction](/spec-content/objects/Restriction.md) object `units` property to `unit`.
- Add values `parking` and `median` to the [LaneType](/spec-content/enumerated-types/LaneType.md) enumerated type.
- Define a new data feed, the [RoadRestrictionFeed](/spec-content/objects/RoadRestrictionFeed.md) and [RestrictionRoadEvent](/spec-content/objects/RestrictionRoadEvent.md) to enable providing a feed of restrictions on roadways, such as bridge clearances.
- Define a new data feed, the [SwzDeviceFeed](/spec-content/objects/SwzDeviceFeed.md), to enable equipment vendors and manufacturers to provide high-level information about deployed field devices in work zones. A `SwzDeviceFeed` contains a new feature type, the [FieldDeviceFeature](/spec-content/objects/FieldDeviceFeature.md), which contains information about a specific type of field device. The following field devices are defined in WZDx v4.0:
    - `ArrowBoard`: An electronic, connected arrow board which can display an arrow pattern to direct traffic.
    - `Camera`: A camera device deployed in the field, capable of capturing still images.
    - `DynamicMessageSign`: An electronic traffic sign deployed on the roadway, used to provide information to travelers.
    - `FlashingBeacon`: A flashing beacon light of any form (e.g. trailer-mounted, vehicle), used to indicate something or capture driver attention.
    - `HybridSign`: A hybrid sign that contains static text (e.g. on an aluminum sign) along with a single electronic message display, used to provide information to travelers.
    - `LocationMarker`: Describes any GPS-enabled ITS device that is placed at a point on a roadway to dynamically know the location of something (often the beginning or end of a work zone).
    - `TrafficSensor`: A traffic sensor deployed on a roadway which captures traffic metrics (e.g. speed, volume, occupancy) over a collection interval.
- Rename the `workers_present` property on the [WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md) object to  `worker_presence`; change the type from "boolean" to a new [WorkerPresence](/spec-content/objects/WorkerPresence.md) object which enables providing more nuanced information about worker presence in work zones.

## Refactoring
- Separate the v3.1 RoadEvent object into [RoadEventCoreDetails](/spec-content/objects/RoadEventCoreDetails.md) (details that are shared by all specific types of road events) and specific types of road events ([WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md), [DetourRoadEvent](/spec-content/objects/DetourRoadEvent.md), and [RestrictionRoadEvent](/spec-content/objects/RestrictionRoadEvent.md)) which each contain the `RoadEventCoreDetails` via a `core_details` property; update the [RoadEventFeature](/spec-content/objects/RoadEventFeature.md) `properties` property to be one of the specific road events types.
- Move the `location_method` property from the [FeedDataSource](/spec-content/objects/FeedDataSource.md) object to the [WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md) object.
- Change the `reduced_speed_limit` property on the [WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md) to `reduced_speed_limit_kph`; change its type from "integer" to "number" and clarify that the value should be in kilometers per hour.
- Deprecate the `lane_number` property on the [Lane](/spec-content/objects/Lane.md) object.
- Deprecate the `lrs_type` and `lrs_url` properties on the [FeedDataSource](/spec-content/objects/FeedDataSource.md) object.
- Remove the deprecated value `alternating-one-way` from the [LaneStatus](/spec-content/enumerated-types/LaneStatus.md) enumerated type.
- Remove the following deprecated properties from the road event (RoadEvent in previous versions; [WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md) and [RoadEventCoreDetails](/spec-content/objects/RoadEventCoreDetails.md) in 4.0):
    - `road_event_id`
    - `road_number`
    - `road_name`
    - `total_num_lanes`
- Remove the following deprecated values from the [LaneType](/spec-content/enumerated-types/LaneType.md) enumerated type: 
    - `left-lane`
    - `right-lane`
    - `middle-lane`
    - `center-lane`
    - `right-shoulder`
    - `left-shoulder`
    - `right-second-exit-ramp`
    - `left-second-exit-ramp`
    - `right-entrance-exit-ramp`
    - `left-entrance-exit-ramp`
    - `hov-lane`
    - `alternating-flow-lane`
    - `reversible-lane`
    - `right-entrance-lane`
    - `left-entrance-lane`
    - `left-entrance-ramp`
    - `right-merging-lane`
    - `left-merging-lane`
    - `right-second-entrance-ramp`
    - `left-second-entrance-ramp`
- Require the `road_names` property on the [RoadEventCoreDetails](/spec-content/objects/RoadEventCoreDetails.md).
- Require the `id` property on the [RoadEventFeature](/spec-content/objects/RoadEventFeature.md).
- Refine the [LaneType](/spec-content/enumerated-types/LaneType.md) enumerated type; specifically:
    - Rename `lane` to `general`.
    - Remove `right-turning-lane` and `left-turning-lane`.
    - Remove `right-exit-lane` and `left-exit-lane`.
    - Add `exit-lane`.
    - Remove `right-exit-ramp` and `left-exit-ramp`.
    - Add `exit-ramp`.
    - Remove `right-entrance-ramp` and `left-exit-ramp`.
    - Add `entrance-ramp`.
    - Add `entrance-lane`.
- Deprecate the `location_verify_method` property on the [FeedDataSource](/spec-content/objects/FeedDataSource.md).
- Update the [SpatialVerification](/spec-content/enumerated-types/SpatialVerification.md) enumerated type value descriptions to clarify that verified work zone locations should use a GPS enabled device.

j-d-b commented:
I added release notes—note that the content of the `Release Notes` section in the project README is identical to the v4.0 section in RELEASES.md and the first PR comment above.

I was thinking we should slim down to the `Release Notes` section of the project README and leave the detailed notes in the RELEASES.md, what do you think?

j-d-b commented:
> I agree, especially considering how long this version's notes are.

I revised the Release Notes section of the project README to remove some specifics (such as the sub-lists).

chuehlien commented on Creating_a_WZDx_Feed.md:
```diff
@@ -0,0 +1,57 @@
+3. The preferred [GeoJSON Geometry](https://tools.ietf.org/html/rfc7946#page-7) for a [RoadEventFeature](/spec-content/objects/RoadEventFeature.md) is `LineString`, which allows indicating the full path of the road event. In cases where only the beginning and ending coordinates are available, the `MultiPoint` can be used.  
+4. A cascading multi-lane closure should consider the speed of vehicles traveling through the work zone. If the distance between lane closures is short enough that time in a to-be-closed lane is not significant, which is common, the to-be-closed lane should be represented as closed to avoid traffic delays and potential crashes.
+5. A detour must be represented with a [RoadEvent](/spec-content/objects/RoadEvent.md) of type (i.e. with `event_type` of) `detour` and its containing [RoadEventFeature](/spec-content/objects/RoadEventFeature.md) should have a `geometry` of `type` `LineString` to represent the full detour route. The detour road event should be connected to the work zone road event using the `relationship` propert ...
+6. If the `lanes` property on the [RoadEvent](/spec-content/objects/RoadEvent.md) object is provided, it must include one entry for every lane in the road event. Providing lane information for only some of the lanes in a road event is not allowed.
```
5 and 6 need to be rephrased to since RoadEvent has been separated out into RoadEventCoreDetails and the specific types of road events.

j-d-b replied:
Good notes, I fixed 6. 6 I think we should remove, because its not really a business rule, just a helpful tip on how to use WZDx.

j-d-b replied:
All, do we want to require detour to have a geometry type of LineString? We can do this via the schema then delete this rule.

mark-mockett replied:
@j-d-b agreed about 5 but this could also be addressed through the larger business rule update. Don't think it needs to be done now. Requiring detour geometry be a linestring is a meaningful change to the spec that doesn't really accomplish much (LineString and Multipoint have essentially the same functionality, in my opinion)

chuehlien commented on spec-content/objects/WorkerPresence.md:
```diff
@@ -0,0 +1,16 @@
+`are_workers_present` | Boolean | Whether workers are present in the work zone event area. This value should be set in accordance with the definition provided in the `definition` property if it is provided. | Required | 
+`definition` | Array; \[[WorkerPresenceDefinition](/spec-content/enumerated-types/WorkerPresenceDefinition.md)\] | A list of situations in which workers are considered to be present in the jurisdiction of the data provider. | Optional
+`method` | [WorkerPresenceMethod](/spec-content/enumerated-types/WorkerPresenceMethod.md) | Describes the method for how worker presence in a work zone event area is determined. | Optional | 
+`worker_presence_last_confirmed_date` | String; [date-time](https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-7.3.1) | Date and time at which the presence of workers was last confirmed. | Optional | All datetime formats shall follow [RFC 3339 Section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6).
```
Do we want to specify if this is UTC (or not)?

e.g. `The UTC time and date when the presence of workers was last confirmed.`

sknick-iastate replied:
I agree it should be for consistency with the `start_date` and `end_date`

j-d-b replied:
Done

chuehlien commented on spec-content/enumerated-types/SpatialVerification.md:
```diff
@@ -4,11 +4,11 @@ An indication of how a geographical coordinate was defined.
-`estimated` | Estimated location associated with work zone activities and lane closures.<br>An estimated measurement may be based on an approximation of a location<br>referencing method (e.g., lat/long or milepost), for example: a point relative to a<br>posted milemarker, point on a map, or GPS device that provides less than<br>centimeter accuracy.
-`verified` | Actual reported information about work zone locations. Actual location is<br>typically measured by a calibrated navigation or survey system to centimeter<br>accuracy (six decimal places for latitude and longitude).
+`estimated` | Estimated location for the work zone road event geometry.  An estimated measurement is based on an approximation of the reported location of a work zone.  Approximations of the location can include but are not limited to a point relative to a posted mile maker or cross street, selecting a point on a map, or locations based on project plans.  
+`verified` | Verified locations for the work zone road event geometry representing the actual extents of the work zone.  A verified measurement is based on actual reported data from a GPS equipped device showing the measured location of the work zone.  Ths actual location is measured using a GPS device that provides the verified location of the identified location.
```
Wondering if the second and third sentence for `verified` description can be combined.

j-d-b replied:
I removed the third sentence.

chuehlien commented:
Added a couple of comments. Please double check my [changes](https://github.com/usdot-jpo-ode/wzdx/compare/ab934f2..e2d6482), since I'm not 100% sure about some of the ones I've made relating to the field device that referenced RoadEvents. I've only reviewed the RoadRestrictionFeed and WZDxFeed contents and have not reviewed the json schema, example files, and only skimmed the example README - hope others are able to provide coverage for those. Good work, all!

j-d-b commented:
> Added a couple of comments. Please double check my [changes](https://github.com/usdot-jpo-ode/wzdx/compare/ab934f2..e2d6482), since I'm not 100% sure about some of the ones I've made relating to the field device that referenced RoadEvents.

@chuehlien your review was excellent and fixed a lot of issues. Thank you so much. I updated two of your edits regarding the use of the TrafficSensorLaneData (this object's tie to the road event is a bit confusing).

If you can make any time to review the schemas at any point, even if it's after the release, I'd really appreciate it as I bet you would find some issues given what you found in the markdown files.

Eventually, I think it would be ideal if the JSON schema is all that we maintain and we can auto-generate more human-friendly docs from it.

j-d-b commented:
> should the WZDx feed also be changed to `feed_info`?

I would like that, but I brought that up at a spec update co-chairs meeting (near the end of the meeting) and at that time the consensus was (which I understood) that it was another breaking change for minimal reason.

However seeing that we don't want to do another major release anytime soon and the change would great greater consistency and less stuff that is only that way for historical reasons, I think we should do it. It's a trivial from a development perspective.

@mark-mockett, @sknick-iastate I think if you both are for it (renaming the WZDxFeed `road_event_feed_info` property to `feed_info` for consistency with the RoadRestrictionFeed and SwzDeviceFeed), let's do it.

sknick-iastate commented:
> > should the WZDx feed also be changed to `feed_info`?
> 
> I would like that, but I brought that up at a spec update co-chairs meeting (near the end of the meeting) and at that time the consensus was (which I understood) that it was another breaking change for minimal reason.
> 
> However seeing that we don't want to do another major release anytime soon and the change would great greater consistency and less stuff that is only that way for historical reasons, I think we should do it. It's a trivial from a development perspective.
> 
> @mark-mockett, @sknick-iastate I think if you both are for it (renaming the WZDxFeed `road_event_feed_info` property to `feed_info` for consistency with the RoadRestrictionFeed and SwzDeviceFeed), let's do it.

Just throwing it out there but what are your thoughts on changing the restriction to `road_event_feed_info`?  Then also changing the SwzDevice to `swz_device_feed_info`?  I know it breaks the consistency but really these would not be in the same feed since it is different producers/consumers.

j-d-b commented:
> Just throwing it out there but what are your thoughts on changing the restriction to `road_event_feed_info`? Then also changing the SwzDevice to `swz_device_feed_info`? 

Since the properties use the same object (the FeedInfo object), I think it is more clear to have the property name be the same for all feeds.

mark-mockett commented:
@j-d-b @sknick-iastate I agree that it'd be cleaner to have each feed info property called the same thing. But I'm also reluctant to make several last-minute revisions that don't improve functionality of the spec (applies to this requiring LineString geometry for detours)

mark-mockett commented:
@j-d-b I need some help understanding what I (accidentally) did with the "Merge branch v4.0 release" commit and its reversion - I switched over to the desktop version of GitHub, and it looks like the first commit just redid all the changes that have been made since the last merge (even though they were already committed). I reverted it but now all those changes are gone. Do I revert the reversion??

sknick-iastate commented:
@j-d-b should the `road_event_id` be removed from the [MarkedLocation Object](https://github.com/usdot-jpo-ode/wzdx/blob/v4.0-release/spec-content/objects/MarkedLocation.md)? It seems redundant since the [FieldDeviceCoreDetails](https://github.com/usdot-jpo-ode/wzdx/blob/v4.0-release/spec-content/objects/FieldDeviceCoreDetails.md) already includes the road event id's. I can't think of a scenario where this information would be different but wanted to verify in case I was overlooking something.

j-d-b commented:
> I agree that it'd be cleaner to have each feed info property called the same thing. But I'm also reluctant to make several last-minute revisions that don't improve functionality of the spec (applies to this requiring LineString geometry for detours)

I think the only revisions we should make right now are those that don't change the functionality of the spec! As for the specific proposal of renaming `road_event_feed_info` to `feed_info`, I'll make an issue for it to address next cycle.

> should the `road_event_id` be removed from the [MarkedLocation Object](https://github.com/usdot-jpo-ode/wzdx/blob/v4.0-release/spec-content/objects/MarkedLocation.md)? 

Good thought and I'd like to continue the discussion but I think this is too major of a change to make now. Can you make an issue for this?

sknick-iastate commented:
@j-d-b sounds good on the MarkedLocation issue.  I have a couple others written out as well but was waiting for v4.0 release to be merged so I can link to the relevant information directly.

j-d-b commented:
> I need some help understanding what I (accidentally) did with the "Merge branch v4.0 release" commit and its reversion...

@mark-mockett I see you reverted the revert commit. That was the right call. It looks like the merge commit was just catching up/combining your local copy to what was new on remote since last time you were working locally.

mark-mockett commented on spec-content/README.md:
```diff
@@ -21,63 +24,104 @@ Each object is described by a table of properties, which are each defined by the
-The object diagram below indicates the relationship between the data objects.
+#### WZDxFeed
 
 ![Object Diagram](/images/wzdx_object_diagram.jpg)
```
We were going to remove this until it can be updated, right?

j-d-b replied:
Yes—replaced with [TBD].

mark-mockett commented on examples/WZDxFeed/scenario3_shoulder_bidirectional_linestring_example.geojson:
```diff
@@ -28,17 +24,22 @@
+          "event_type": "work-zone",
+          "relationship": {
+            "parents": [
+              "121388"
```
Might not be necessary for this release if we expand examples soon, but there should be a WorkZoneRoadEvent in this feed with the ID 121388?

j-d-b replied:
There doesn't need to be. That ID could correspond to the project or a non-WZDx entity. It is a confusing system. See #197.

mark-mockett reviewed (approved):
A couple small things to resolve but generally everything is looking good.

j-d-b commented on spec-content/enumerated-types/WorkerPresenceMethod.md:
```diff
@@ -0,0 +1,20 @@
+# Worker Presence Method
```
Great, I like your recommendation as it's consistent with the object markdown file titles.

j-d-b replied:
Are you going to work on that now? It doesn't need to happen for the release so you could alternatively create an issue for it and do it whenever, since it's editorial only.

mark-mockett replied:
just changed them

sknick-iastate reviewed (approved):
I was able to review all of the SwzDevice and most of the WZDx.  Looks good to me!
